### PR TITLE
Added cell->is_locally_owned() in project_boundary_values_div_conforming

### PR DESCRIPTION
**Pull request by JonMatthewsFCAS**
I added a check for cells being locally owned in the implementations of project_boundary_values_div_conforming for use with Raviart-Thomas elements. Without this it causes a seg fault when used on a parallel distributed triangulation.

I also added a test program. I'm not sure if this is what's needed for a proper test as I've never written one before. I am open to any suggestions on how to improve upon it.

### CODE REVIEW

File: tests/mpi/project_bv_div_conf.cc [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+      setup_system ();
+      pcout << "System set up." << std::endl;
+    }
+
+} // end namespace ResFlow
+
+int
+main (int argc, char *argv[])
+{
+  const unsigned int dim = 2;
```
Comment (jppelteret) [inline tests/mpi/project_bv_div_conf.cc]: You'll have to add some lines like [this](https://github.com/dealii/dealii/blob/master/tests/mpi/attach_data_02.cc#L202-L203) here:
```c++
Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
MPILogInitAll log;
```

File: tests/mpi/project_bv_div_conf.cc [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    {
+      std::cerr << std::endl << std::endl
+	  << "----------------------------------------------------"
+	  << std::endl;
+      std::cerr << "Unknown exception!" << std::endl << "Aborting!" << std::endl
+	  << "----------------------------------------------------"
+	  << std::endl;
+      return 1;
+    }
+
```
Comment (jppelteret) [inline tests/mpi/project_bv_div_conf.cc]: You could add 
```c++
deallog << "OK" << std::endl;
```
here to indicate that the program ran successfully.

File: tests/mpi/project_bv_div_conf.cc [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_raviart_thomas.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_refinement.h>
+
+#include <iostream>
+
```
Comment (jppelteret) [inline tests/mpi/project_bv_div_conf.cc]: You need to include the test [master header](https://github.com/dealii/dealii/blob/master/tests/mpi/attach_data_02.cc#L32) here:
```c++
#include "../tests.h"
```

**Comment by jppelteret** [comment]
Oh, and you also need to ensure that the indentation is correct. You should run `make indent` before you commit your changes. If the indentation tool is not installed on your machine, then some instructions will be issues that explain how to install it.

**Comment by JonMatthewsFCAS** [comment]
I will get to work on those items. Thanks for your help and patience.

**Comment by JonMatthewsFCAS** [comment]
Okay, I made the suggested changes, but I'm stuck on a couple of things.

1) I tried to output the contents of the `AffineConstraints` object by calling
`constraints.print(deallog);`
but this won't build because `deallog` can't be converted from `dealii::LogStream` to `std::ostream&`. Is there a relatively easy way I can send this output to the `deallog`?

2) `make indent` runs, but it doesn't seem to be changing the files. It gives the output
`Indenting recently changed files in the deal.II directories`
`Built target indent`,
so it looks like it's working.
Is there some syntax I need to add to tell it what files to re-indent? I've touched them and even made extra edits to try to force it to no avail.

3) Once I'm done, I'm not sure how to resubmit here. Do I create a new pull request, or can I push my changes into this pull request somehow? Also, is there some way I can run those travis-ci checks before submitting, so that I can catch the errors and not waste people's time?

**Comment by bangerth** [comment]
On 10/12/2018 01:18 AM, JonMatthewsFCAS wrote:
> I tried to output the contents of the |AffineConstraints| object by calling
> |constraints.print(deallog);|
> but this won't build because |deallog| can't be converted from 
> |dealii::LogStream| to |std::ostream&|. Is there a relatively easy way I 
> can send this output to the |deallog|?

You can do
   deallog.get_file_stream()
to get a std::ostream object into which you can write.

**Comment by jppelteret** [comment]
>I tried to output the contents of the AffineConstraints object by calling
constraints.print(deallog);
but this won't build because deallog can't be converted from dealii::LogStream to std::ostream&. Is there a relatively easy way I can send this output to the deallog?

@bangerth covered this in comment above

> make indent runs, but didn't changing the files. It gives the output
Indenting recently changed files in the deal.II directories
Built target indent,
so it looks like it's working.

If there are no changes to the indentation after running `make indent`, then (assuming that all of your other changes had been committed) a quick call to `git status` will verify that there are no files that were modified by the indenter.

> It looks like I need to run indent from the top level directory of deal.ii rather than my build folder to make it work correct? 

You should be able to run this from the build directory. Otherwise this command is the equivalent that can be called from the top level directory:
```sh
$ ./contrib/utilities/indent
```

> When I ran it on my test, I got an error: indent: tests/mpi/project_bv_div_conf.cc:147: Error:Unmatched 'else', but it looks like that else is matched with an if, so I'm not sure what the problem is.

Hmm... I don't know what the issue could be. I don't see any problems with the code block between lines 145 and 154.

> It looks like if I push any commits to my clone of dealii, it flows through to here. Is there some way I can run those travis-ci checks locally before pushing my changes, so that I can catch the errors and not waste people's time?

Well, that's the idea behind running `make indent` before pushing anything upstream. Similar for re-checking your changes to both the library and the test itself using `ctest`. But its hardly an issue if you push a new commit and there's an error. When the PR is nearly done, then we'll issue a command to run the tester, which will actually execute this test (and use more compute time).

As a further tip, here's a link to the documentation on [adding a test to the testsuite](https://dealii.org/developer/developers/testsuite.html#layoutaddtests). Sorry that I forgot to direct you to this earlier.

Also, if you feel uncomfortable rebasing your branch, dropping commits, etc. then remember that you can always make a backup of your branch by checking out a new branch with this branch as the source. If you feel like you're completely stuck then let us know and one of us could try to help you fix things (but it would be a good exercise to try to work this out, if you're willing to do so ;-)

**Comment by jppelteret** [comment]
@JonMatthewsFCAS Whoops, I myself forgot to run `make indent` one last time before submitting those patches to you. May you please make the following (whitespace) changes to the test in order to finally make the indenter happy?
```diff
-      const unsigned int               press_order = 1;
-      ResFlowProblem<dim>              resflow(press_order);
+      const unsigned int  press_order = 1;
+      ResFlowProblem<dim> resflow(press_order);
```

**Comment by jppelteret** [comment]
> drop/squash the commits into something that makes a bit more sense

Yes, in this case this is certainly desirable. If you'd like me to help you do it then I think that you need to select the option in Github that [gives maintainers permission to make changes to your branch](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/). Otherwise you should read a couple of tutorials on how to do this (google something like "git squash" and "git rebase interactive"). 

Also... this has been quite a lot of work, and you deserve some credit for it! Feel free to add an entry describing the bug fix to `./doc/news/changes/minor`. You should take one of the other entries as a template for this entry to the changelog.

**Comment by masterleinad** [comment]
> Yes, in this case this is certainly desirable. If you'd like me to help you do it then I think that you need to select the option in Github that [gives maintainers permission to make changes to your branch](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/). Otherwise you should read a couple of tutorials on how to do this (google something like "git squash" and "git rebase interactive").

@jppelteret You already pushed to the branch, so you should be able to squash the commits as well.

**Comment by jppelteret** [comment]
> You already pushed to the branch, so you should be able to squash the commits as well.

No I didn't. I opened a PR ;-)

**Comment by JonMatthewsFCAS** [comment]
I have "Allow edits from maintainers." checked, so I think you can squash the commits if you want to wrap this up soon. 

I'm happy to work on it myself, but I have to go to my part time job for the next five hours and won't have access to a computer during that time to work on it until afterward. If it's not done when I get back, I'll jump right on it. 

Thanks for all the help,
Jon

**Comment by jppelteret** [comment]
>While there, could you please switch to range loop in the three places:

Where?

**Comment by jppelteret** [comment]
> Where?

I've spotted which lines you're referring to. This doesn't make much sense, since there are many other lines in the same functions, as well as the whole of `vector_tools.templates.h` that don't use the range-based loops. I think that this should be done consistently in another PR.

**Comment by bangerth** [comment]
Let's give a new contributor some slack and accept the patch as is :-)

We really do appreciate all contributions and part of the process of going through submitting your first pull request is that you learn about this whole process works. That might include learning how to do better in the future, but we shouldn't let "perfect" get in the way of "we appreciate what you are doing here" :-)

**Comment by bangerth** [comment]
That said, @JonMatthewsFCAS , in case you want an easy way to add to your number of accepted patches: You can always go through this file and just replace the existing cell loops by range-based ones as a follow-up :-) As I said above, we really do appreciate all contributions, even those that just fix up these little things and modernize the code without actually adding functionality!

**Comment by JonMatthewsFCAS** [comment]
Thanks everyone. I definitely appreciate all the help, and I learned a ton in the process. 

@bangerth I will work on that change you and @davydden suggested. It would be a good idea for me to try another patch while the process is still fresh in my memory.

**Comment by bangerth** [comment]
Excellent, @JonMatthewsFCAS -- we look forward to it!

**Pull request merged**